ykutil/src/obj.rs

```diff
@@ -27,27 +23,24 @@ pub static SELF_BIN_PATH: LazyLock<PathBuf> = LazyLock::new(|| {
     PathBuf::from(unsafe { CStr::from_ptr(info.dli_fname) }.to_str().unwrap())
 });
 
-/// The current executable mmaped into the address space.
-///
-/// PERF: Consider making the segment containing .llvmbc loadable so that ld.so loads it
-/// automatically when spawning the process.
-static EXE_MMAP: LazyLock<Mmap> = LazyLock::new(|| {
-    let pathb = env::current_exe().unwrap();
-    let file = fs::File::open(&pathb.as_path()).unwrap();
-    unsafe { memmap2::Mmap::map(&file).unwrap() }
-});
-
-/// The current executable parsed to an `object::File`.
-static EXE_OBJ: LazyLock<object::File> =
-    LazyLock::new(|| object::File::parse(&**EXE_MMAP).unwrap());
+/// The `llvm.embedded.module` symbol in the `.llvmbc` section.
+#[repr(C)]
+struct EmbeddedModule {
+    /// The length of the bitcode.
+    len: usize,
+    /// The start of the bitcode itself.
+    first_byte_of_bitcode: u8,
```

Comment:
This may be a silly question, but why is this a `u8` and why can we just use a reference to this `u8` later on to load the bitcode from? I assume some magic happens in `#[link_name = "llvm.embedded.module"]` but I'd be curious to know what this magic is.

Reply:
The `.llvmbc` section is now magically loaded as the interpreter starts, and it contains a symbol `llvm.embedded.module` whose binary format we know exactly at runtime:

 - a `usize` `length` field.
 - a bitcode payload of length `length` (the previous field).

(and here's no padding between these)

Thanks to the recent `ykllvm` change, the `extern C` block anchors a Rust-level struct with identical ABI to this symbol (hence the `link_name` bit). This means that when we read this struct in Rust, we actually read the symbol generated by ykllvm.

We don't know the length of the payload statically, but if we did, I'd have used an array instead of a `u8`. Since we don't, the `first_byte_of_bitcode` field just lets us get a reference to the beginning of the bitcode payload.

The LLVM API to load bitcode requires a pointer to the bitcode and the length of it. The length is `len` and the pointer to the bitcode is `&first_byte_of_bitcode`. The memory for the section is static, and thus eternal, so it's safe to pass pointers to it around.

Hope that makes sense.

Reply:
Since there's no padding, I would probably suggest removing `first_byte_of_bitcode` and making the calculation of that address `std::mem::sizeof::<usize>()`, but I don't feel strongly about it.

Reply:
That's what I started with, but changed to just using the ABI of the struct, since it requires fewer imports and unsafe blocks. I prefer it this way.

Reply:
OK.